### PR TITLE
add publishing to typescript sdk

### PR DESCRIPTION
**Pull request by ncitron**
replacing #2983 

@666lcz let me know if there are any issues when testing this on main. I've also added an example of how to use the new feature in the readme.

### CODE REVIEW

File: sdk/typescript/README.md [hunk 1]
```diff
@@ -151,3 +151,24 @@ const moveCallTxn = await signer.executeMoveCall({
...
+// Generate a new Keypair
+const keypair = new Ed25519Keypair();
+const signer = new RawSigner(
+  keypair,
+  new JsonRpcProvider('https://gateway.devnet.sui.io:443')
+);
+const bytecode = await fs.readFile('path/to/project/build/project_name/bytecode_modules/module_name.mv', 'base64');  
+const publishTxn = await signer.publish(
+  {
+    packages: [bytecode.toString()],
```
Comment (666lcz) [inline sdk/typescript/README.md:169]:
```suggestion
    compiledModules: [bytecode.toString()],
```

ditto

**Comment by damirka** [comment]
@666lcz 

> Another thing I noticed that is it's a little bit tricky to publish packages that have dependencies or multiple modules. @damirka I remember that you also experimented with publishing on the TS side, any suggestion here?

We just stack everything together and they're published as a single package. Nothing too hard. It's even better to do it this way, so we can use Move's functionality such as `public(friend)` - which can only exist in multi-module publish.

**Review by damirka — APPROVED** [review]

**Review by Clay-Mysten — APPROVED** [review]
Thanks, Damir! If they need to modify the `path/to/project/build/` path in the publishing example, I would tell them that before the example. Otherwise, LGTM. Thanks!

**Comment by ncitron** [comment]
Great! I've committed the suggestions and resolved some merge conflicts.

File: sdk/typescript/src/signers/txn-data-serializers/rpc-txn-data-serializer.ts [hunk 2]
```diff
@@ -116,4 +117,25 @@ export class RpcTxnDataSerializer implements TxnDataSerializer {
...
+  async newPublish(
+    signerAddress: SuiAddress,
+    t: PublishTransaction
+  ): Promise<Base64DataBuffer> {
+    try {
+      const resp = await this.client.requestWithType(
+        'sui_publish',
+        [
+          signerAddress,
+          t.packages,
```
Comment (666lcz) [inline sdk/typescript/src/signers/txn-data-serializers/rpc-txn-data-serializer.ts:130]: this should be `t.compiledModules`

Comment (ncitron) [reply sdk/typescript/src/signers/txn-data-serializers/rpc-txn-data-serializer.ts]: Good catch. Just pushed the change

**Pull request merged**